FCMP++ consensus + CLI/wallet2 integration

j-berman opened the pull request:
Ripe for Carrot integration :)

j-berman committed:
Fix sum of inputs == sum of outputs

j-berman committed:
FCMP++: batch verification in conensus

j-berman committed:
fcmp++: sort fcmp_pp_params inputs alongside key image sort

j-berman committed:
fcmp++: fix balace_last_pseudo_out, don't consume inputs

j-berman committed:
fcmp++: sweep uses FCMP++ max inputs

jeffro256 commented:
Merged/rebased in commits 20a374350392f8870d1f41ae0633180efbc5091a to 80e81669b3223509be0ce50745bc6d9ae1db518d

j-berman committed:
Fix compile error from cherry-pick

j-berman committed:
fcmp++ small db touchups

j-berman committed:
fcmp++: default init FCMP++ tx rctSig prunable fields

j-berman committed:
fcmp++: fix JSON serialization

j-berman committed:
fcmp++: input and output limits in tx wallet2 tx constr

j-berman committed:
fcmp++: wallet2 use new transfer unlock logic at fork

j-berman committed:
functional tests: test spend right at unlock

j-berman committed:
functional tests: sweep_all to transfer_split in dbl spend prtction

Since FCMP++ imposes stricter input/output limits, sweep_all can
end up taking multiple txs. Using transfer_split keeps the spirit
of the test while using a single tx as expected.

j-berman committed:
functional tests: new test for sweep_all

j-berman committed:
fcmp++: fix batch verification short-circut

j-berman committed:
Update wallet API for ring size handling

j-berman committed:
fcmp++: trim to cached tree edge, instead of re-calculating edge

Suggestion 1 by @jeffro256 here:
https://github.com/monero-project/monero/pull/9436#issuecomment-2519858103

j-berman committed:
functional tests: test for daemon reorg recovery handling

j-berman committed:
fcmp++: delay trim_tree until removing block containing tree outputs

Suggestion 2 by @jeffro256 here:
https://github.com/monero-project/monero/pull/9436#issuecomment-2519858103

I set up a new table for tree meta (stores n block tuples and
tree root), and stopped using block info. I did this because
popping a block removes the block info, but we still need to know
tree state data because of this change, to know how to trim the
tree. Technically don't need both tree edge and tree root stored,
so there is redundancy there. But this approach seemed cleaner.

j-berman committed:
db: Don't store tree root in addition to tree edge

j-berman committed:
db: store tree edge as group, don't need layer_idx

j-berman committed:
Fix curve_trees db tests via alterations to grow_tree

j-berman committed:
fcmp++: fix tree cache reorg handling after syncing n blocks

Since we sometimes sync multiple blocks at a time, and syncing
is optimized by constructing a single tree extension, in order to
trim to a prior tree edge correctly, we need to make sure we're
using the correct version of the prior tree edge. We can either do
more hashes when forward syncing to calculate each block's tree
edge, or we can calculate the expected tree edge on reorg. This
commit opts for the latter to optimize forward sync.

j-berman committed:
fucntional tests: test for wallet reorg recovery handling

j-berman committed:
fcmp++: temp use my fcmp-plus-plus crate fork

j-berman committed:
fcmp++: some cleanup in tree_cache

- Moved more division logic / index calculation into the curve
trees class and re-used that.

j-berman committed:
Rename OutputsByLastLockedBlock -> OutsByLastLockedBlock
